src/wallet/spend.cpp

```diff
@@ -620,15 +630,59 @@ static void DiscourageFeeSniping(CMutableTransaction& tx, FastRandomContext& rng
     // enough, that fee sniping isn't a problem yet, but by implementing a fix
     // now we ensure code won't be written that makes assumptions about
     // nLockTime that preclude a fix later.
+    //
+    // nSequence-based
+    // ---------------
+    //
+    // If all inputs in the tx spend taproot coins, use nSequence to discourage
+    // fee-sniping. This creates a "privacy cloak" for txs that use taproot
+    // key-path spends (BIP326).
+    //
+    // Compared to nLockTime-based anti-fee-sniping, this may be minimally less
+    // effective, as nSequence only allows relative locks. If all inputs of the
+    // tx can be moved backward in the chain, this tx can be as well. Though,
+    // this should be fine given that 10% of txs have their locktime randomly
+    // reduced to a past block. Also, sequence-based locking will only be used
+    // if every sequence is at least 1.
     if (IsCurrentForAntiFeeSniping(chain, block_hash)) {
-        tx.nLockTime = block_height;
-
-        // Secondly occasionally randomly pick a nLockTime even further back, so
-        // that transactions that are delayed after signing for whatever reason,
-        // e.g. high-latency mix networks and some CoinJoin implementations, have
-        // better privacy.
-        if (rng_fast.randrange(10) == 0) {
-            tx.nLockTime = std::max(0, int(tx.nLockTime) - int(rng_fast.randrange(100)));
+        auto locktime_based{[&] {
+            if (!SignalsOptInRBF(CTransaction{tx})) {
+                // Use locktime if the tx inputs do not opt-in to RBF
+                return true;
+            }
+            for (const auto& in : inputs) {
+                if (in.depth < 1 || uint64_t(in.depth) > MAX_BLOCK_SEQ) {
+                    // Use locktime if any depth is out-of-range
+                    return true;
+                }
+                std::vector<std::vector<uint8_t>> dummy;
+                const TxoutType in_type{Solver(in.txout.scriptPubKey, dummy)};
+                // Use locktime if any input isn't taproot
+                if (in_type != TxoutType::WITNESS_V1_TAPROOT) {
+                    return true;
+                }
+            }
+            return false;
+        }};
+        if (locktime_based() || rng_fast.randrange(2) == 0) {
```

Comment:
nit:

Use `randbool`

```suggestion
        if (locktime_based() || rng_fast.randbool()) {
```

Also, why not have the random bool be part of `locktime_based`?

Reply:
Yeah, it can be moved inside, as all early returns are `true` and the final return is the only one with `false`.